app/components/post_draft/draft_handler/draft_handler.tsx

```diff
@@ -0,0 +1,321 @@
+        if (uploadErrorTimeout.current) {
+            clearTimeout(uploadErrorTimeout.current);
+        }
+        setUploadError(error);
+
+        uploadErrorTimeout.current = setTimeout(() => {
+            setUploadError(null);
+        }, UPLOAD_ERROR_SHOW_INTERVAL);
+    }, []);
+
+    const addFiles = useCallback((newFiles: FileInfo[]) => {
+        if (!newFiles.length) {
+            return;
+        }
+
+        if (!canUploadFiles) {
+            newUpdateError(uploadDisabledWarning(intl));
+            return;
+        }
+
+        const availableCount = maxFileCount - currentFiles.length;
+        if (newFiles.length > availableCount) {
+            newUpdateError(fileMaxWarning(intl, maxFileCount));
+            return;
+        }
+
+        const largeFile = newFiles.find((file) => file.size > maxFileSize);
+        if (largeFile) {
+            newUpdateError(fileSizeWarning(intl, maxFileSize));
+            return;
+        }
+
+        for (const file of newFiles) {
+            prepareUpload(file, fileHandler, serverUrl, channelId);
+        }
+
+        setUploading(true);
+        newUpdateError(null);
+
+        const newFileList = [...currentFiles, ...newFiles];
+        updateFiles(newFileList);
+    }, [intl, maxFileCount, maxFileSize, serverUrl, channelId, currentFiles, updateFiles]);
+
+    const removeFile = useCallback((file: FileInfo) => {
+        const handler = fileHandler.current[file.clientId!];
+        if (handler) {
+            handler.cancel?.();
+        }
+
+        const newFileList = currentFiles.filter((f) => f.clientId !== file.clientId);
+        updateFiles(newFileList);
+    }, [currentFiles, updateFiles]);
+
+    const retryFileUpload = useCallback((file: FileInfo) => {
+        const index = currentFiles.findIndex((f) => f.clientId === file.clientId);
+        if (index == null) {
+            return;
+        }
+
+        const newFile = {...file};
+        newFile.loading = true;
+        newFile.progress = 0;
+        newFile.failed = false;
+
+        prepareUpload(newFile, fileHandler, serverUrl, channelId);
+        setUploading(true);
+        updateFiles([...currentFiles.slice(0, index), newFile, ...currentFiles.slice(index + 1)]);
+    }, [serverUrl, channelId, currentFiles, updateFiles]);
+
+    useEffect(() => {
+        return () => {
+            if (uploadInterval.current) {
+                clearInterval(uploadInterval.current);
+            }
+            if (uploadErrorTimeout.current) {
+                clearTimeout(uploadErrorTimeout.current);
+            }
+        };
+    }, []);
+
+    useEffect(() => {
+        if (uploading && !uploadInterval.current) {
+            uploadInterval.current = setInterval(() => handlerJob.current(), UPLOAD_INTERVAL);
+        }
+
+        if (!uploading && uploadInterval.current) {
+            clearTimeout(uploadInterval.current);
+            uploadInterval.current = undefined;
+        }
+    }, [uploading]);
+
+    useEffect(() => {
```

Comment:
All this works fine for file upload if we stay in the channel while the file gets uploaded. The problem is the moment we switch channels or send the app to the background. What is the expected behaviour on those? Any ideas on how to handle that beautifully?

Reply:
We could move the fileHandler from the out of the component and get it into some ephemeral store. Then the component, on load, will have to check if the files are there, and what is the state, and updated the draft accordingly.

Thoughts @enahum ?